File: packages/billing-entitlements/src/service.ts
import type { BillingConfiguration, BillingRepository, Clock, MembershipPort, OrganizationEntitlement, OrganizationSubscription, SeatSnapshot, SubscriptionQuantityPort, SubscriptionStatus, VerifiedBillingEvent } from "./contracts.ts";

const statusOf=(value:unknown):SubscriptionStatus => {
  const allowed:SubscriptionStatus[]=["incomplete","trialing","active","past_due","unpaid","canceled","paused"];
  return allowed.includes(value as SubscriptionStatus)?value as SubscriptionStatus:"incomplete";
};
const stringField=(o:Readonly<Record<string,unknown>>,key:string)=>typeof o[key]==="string"?o[key] as string:undefined;
const numberField=(o:Readonly<Record<string,unknown>>,key:string)=>typeof o[key]==="number"?o[key] as number:undefined;
function organizationId(object:Readonly<Record<string,unknown>>):string|undefined {
  const metadata=object.metadata;
  return metadata&&typeof metadata==="object"?stringField(metadata as Readonly<Record<string,unknown>>,"organization_id"):undefined;
}
export function deriveEntitlement(subscription:OrganizationSubscription|undefined,now:Date,graceSeconds:number):OrganizationEntitlement {
  const verified_at=now.toISOString();
  if(!subscription)return {organization_id:"unknown",key:"workspace_access",state:"disabled",reason:"no_verified_subscription",verified_at};
  if(subscription.status==="active"||subscription.status==="trialing")return {organization_id:subscription.organization_id,key:"workspace_access",state:"enabled",source_subscription_id:subscription.subscription_id,reason:`subscription_${subscription.status}`,verified_at};
  if(subscription.status==="past_due"&&graceSeconds>0&&subscription.current_period_end&&now.getTime()<=new Date(subscription.current_period_end).getTime()+graceSeconds*1000)return {organization_id:subscription.organization_id,key:"workspace_access",state:"grace",source_subscription_id:subscription.subscription_id,reason:"payment_past_due_grace",verified_at};
  return {organization_id:subscription.organization_id,key:"workspace_access",state:"disabled",source_subscription_id:subscription.subscription_id,reason:`subscription_${subscription.status}`,verified_at};
}
export class BillingEntitlementService {
  private readonly repository:BillingRepository;
  private readonly clock:Clock;
  private readonly configuration:BillingConfiguration;
  constructor(repository:BillingRepository,clock:Clock,configuration:BillingConfiguration){this.repository=repository;this.clock=clock;this.configuration=configuration;}
  async consumeVerifiedEvent(event:VerifiedBillingEvent):Promise<"applied"|"duplicate"|"ignored_stale"|"ignored_unmapped"> {
    if(await this.repository.beginEvent(event)==="duplicate")return "duplicate";
    try {
      if(!event.type.startsWith("customer.subscription.")){await this.repository.completeEvent(event.id);return "ignored_unmapped";}
      const org=organizationId(event.object);const customer=stringField(event.object,"customer");const id=stringField(event.object,"id");
      if(!org||!customer||!id)throw new Error("Subscription webhook is missing organization metadata or provider identifiers.");
      const current=await this.repository.getSubscription(org);
      if(current&&current.provider_event_created>event.created){await this.repository.completeEvent(event.id);return "ignored_stale";}
      const quantity=Math.max(0,numberField(event.object,"quantity")??numberField(event.object,"items_quantity")??current?.quantity??0);
      const periodEnd=numberField(event.object,"current_period_end");
      const subscription:OrganizationSubscription={organization_id:org,customer_id:customer,subscription_id:id,status:event.type.endsWith(".deleted")?"canceled":statusOf(event.object.status),quantity,provider_event_created:event.created,current_period_end:periodEnd?new Date(periodEnd*1000).toISOString():undefined,cancel_at_period_end:event.object.cancel_at_period_end===true};
      await this.repository.saveSubscription(subscription);
      await this.repository.saveEntitlement(deriveEntitlement(subscription,this.clock.now(),this.configuration.past_due_grace_seconds));
      await this.repository.completeEvent(event.id);return "applied";
    } catch(error){await this.repository.failEvent(event.id,error instanceof Error?error.message:"unknown_error");throw error;}
  }
}
export async function reconcileSeats(input:{organization_id:string;subscription:OrganizationSubscription;memberships:MembershipPort;quantities:SubscriptionQuantityPort;repository:BillingRepository;clock:Clock;apply:boolean}):Promise<SeatSnapshot>{
  const billable_members=await input.memberships.countBillableMembers(input.organization_id);
  const snapshot={organization_id:input.organization_id,billable_members,subscription_quantity:input.subscription.quantity,observed_at:input.clock.now().toISOString(),matches:billable_members===input.subscription.quantity};
  await input.repository.saveSeatSnapshot(snapshot);
  if(!snapshot.matches&&input.apply)await input.quantities.setQuantity(input.subscription.subscription_id,billable_members,`seat-reconcile:${input.organization_id}:${snapshot.observed_at.slice(0,10)}`);
  return snapshot;
}

File: backend/src/billing_v2/service.py
from __future__ import annotations

from .models import EntitlementDecision, SubscriptionRecord, SubscriptionStatus, VerifiedStripeEvent
from .ports import BillingRepository, TelemetryPort

VALID_STATUSES: set[str] = {"incomplete", "trialing", "active", "past_due", "unpaid", "canceled", "paused"}


def derive_entitlement(subscription: SubscriptionRecord | None) -> EntitlementDecision:
    if subscription is None:
        return EntitlementDecision("unknown", "disabled", "no_verified_subscription", None)
    enabled = subscription.status in {"active", "trialing"}
    return EntitlementDecision(subscription.organization_id, "enabled" if enabled else "disabled", f"subscription_{subscription.status}", subscription.subscription_id)


class BillingWebhookProcessor:
    def __init__(self, repository: BillingRepository, telemetry: TelemetryPort):
        self.repository = repository
        self.telemetry = telemetry

    async def process(self, event: VerifiedStripeEvent) -> str:
        if await self.repository.begin_event(event) == "duplicate":
            self.telemetry.counter("billing.webhook.duplicate", event_type=event.type)
            return "duplicate"
        try:
            if not event.type.startswith("customer.subscription."):
                await self.repository.complete_event(event.id)
                return "ignored"
            metadata = event.object.get("metadata")
            organization_id = metadata.get("organization_id") if isinstance(metadata, dict) else None
            subscription_id, customer_id = event.object.get("id"), event.object.get("customer")
            if not all(isinstance(value, str) and value for value in (organization_id, subscription_id, customer_id)):
                raise ValueError("Verified subscription event lacks organization metadata or provider IDs.")
            current = await self.repository.get_subscription(organization_id)
            if current and current.provider_event_created > event.created:
                await self.repository.complete_event(event.id)
                self.telemetry.counter("billing.webhook.stale", event_type=event.type)
                return "ignored_stale"
            raw_status = "canceled" if event.type.endswith(".deleted") else event.object.get("status", "incomplete")
            status: SubscriptionStatus = raw_status if raw_status in VALID_STATUSES else "incomplete"  # type: ignore[assignment]
            quantity = event.object.get("quantity", current.quantity if current else 0)
            record = SubscriptionRecord(organization_id, customer_id, subscription_id, status, max(0, int(quantity)), event.created, int(event.object["current_period_end"]) if event.object.get("current_period_end") else None)
            await self.repository.save_subscription(record)
            await self.repository.save_entitlement(derive_entitlement(record))
            await self.repository.complete_event(event.id)
            self.telemetry.counter("billing.webhook.applied", event_type=event.type)
            return "applied"
        except Exception as exc:
            await self.repository.fail_event(event.id, str(exc))
            self.telemetry.error(exc, event_id=event.id, event_type=event.type)
            raise
